Add quality scale configuration for UniFi Protect integration

RaHehl opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

# UniFi Protect Quality Scale: Platinum

This PR adds the quality scale configuration for the UniFi Protect integration, setting it to **bronze** level.

## Bronze
- [x] `action-setup` - Services are registered in `async_setup` via [`async_setup_services(hass)`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/__init__.py#L68)
- [x] `appropriate-polling` - **Exempt**: Integration is push-based using WebSockets (`iot_class: local_push`), no polling needed
- [x] `brands` - UniFi Protect has branding assets in the [brands repository](https://github.com/home-assistant/brands/tree/master/custom_integrations/unifiprotect)
- [x] `common-modules` - Uses [`data.py`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/data.py) for data coordination and [`entity.py`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/entity.py) for base entity class
- [x] `config-flow-test-coverage` - [`test_config_flow.py`](https://github.com/home-assistant/core/blob/dev/tests/components/unifiprotect/test_config_flow.py) provides comprehensive coverage of all config flow paths
- [x] `config-flow` - Full config flow implementation in [`config_flow.py`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/config_flow.py) with VERSION = 2
    - [x] Uses `data_description` to give context to fields
    - [x] Uses `ConfigEntry.data` and `ConfigEntry.options` correctly
- [x] `dependency-transparency` - [uiprotect library](https://pypi.org/project/uiprotect/) is open source on PyPI with tagged releases
- [x] `docs-actions` - All 5 service actions documented at [home-assistant.io/integrations/unifiprotect](https://www.home-assistant.io/integrations/unifiprotect/#action-unikiprotectadd_doorbell_text)
- [x] `docs-high-level-description` - [Documentation](https://www.home-assistant.io/integrations/unifiprotect/) describes UniFi Protect integration
- [x] `docs-installation-instructions` - [Prerequisites section](https://www.home-assistant.io/integrations/unifiprotect/#prerequisites) with Local User setup, Camera Streams, API key creation
- [x] `docs-removal-instructions` - Standard removal process applies, no special steps needed
- [x] `entity-event-setup` - Uses `async_on_remove` pattern and [`entry.async_on_unload`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/__init__.py#L163) for cleanup
- [x] `entity-unique-id` - All entities have unique_id based on device MAC and description key in [`entity.py`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/entity.py#L136)
- [x] `has-entity-name` - [`BaseProtectEntity`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/entity.py#L115) sets `_attr_has_entity_name = True`
- [x] `runtime-data` - Uses `entry.runtime_data` with typed [`UFPConfigEntry = ConfigEntry[ProtectData]`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/data.py#L50)
- [x] `test-before-configure` - Config flow validates connection via [`_async_get_nvr_data`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/config_flow.py#L246) before entry creation
- [x] `test-before-setup` - [`async_setup_entry`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/__init__.py#L78) calls `protect.update()` and raises `ConfigEntryNotReady`/`ConfigEntryAuthFailed`
- [x] `unique-config-entry` - Uses [`async_set_unique_id(nvr_data.mac)`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/config_flow.py#L369) and `_abort_if_unique_id_configured()`

## Silver
- [x] `action-exceptions` - Services raise `HomeAssistantError` and `ServiceValidationError` appropriately in [`services.py`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/services.py)
- [x] `config-entry-unloading` - [`async_unload_entry`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/__init__.py#L168) implemented with platform unloading and `data_service.async_stop()`
- [x] `docs-configuration-parameters` - [Options flow parameters](https://www.home-assistant.io/integrations/unifiprotect/#options) documented (RTSP, realtime metrics, max media)
- [x] `docs-installation-parameters` - Hardware/software requirements and [prerequisites](https://www.home-assistant.io/integrations/unifiprotect/#prerequisites) fully documented
- [x] `entity-unavailable` - WebSocket state handling marks entities unavailable on connection loss in [`data.py`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/data.py#L287)
- [x] `integration-owner` - [@RaHehl](https://github.com/RaHehl) listed as codeowner in [`manifest.json`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/manifest.json#L4)
- [x] `log-when-unavailable` - [`_async_update_change`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/data.py#L269) logs connection loss and restoration
- [x] `parallel-updates` - All platform files set `PARALLEL_UPDATES = 0` for coordinator-based updates (e.g., [`sensor.py`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/sensor.py#L47))
- [x] `reauthentication-flow` - [`async_step_reauth`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/config_flow.py#L290) and `async_step_reauth_confirm` implemented
- [x] `test-coverage` - Test coverage is 99% across all modules, exceeding 95% requirement

## Gold
- [x] `devices` - All entities provide `DeviceInfo` with comprehensive device metadata in [`entity.py`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/entity.py#L118)
- [x] `diagnostics` - [`diagnostics.py`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/diagnostics.py) implements `async_get_config_entry_diagnostics` with anonymized data
- [x] `discovery-update-info` - Discovery updates IP via [`_abort_if_unique_id_configured(updates={CONF_HOST})`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/config_flow.py#L160)
- [x] `discovery` - Supports DHCP, SSDP, and integration_discovery via [unifi-discovery library](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/manifest.json#L7-L45)
- [x] `docs-data-update` - [Documentation](https://www.home-assistant.io/integrations/unifiprotect/) mentions Local Push via WebSockets for real-time updates
- [x] `docs-examples` - Multiple [automation examples](https://www.home-assistant.io/integrations/unifiprotect/#automation-examples) for notifications, doorbell, NFC, fingerprint events
- [x] `docs-known-limitations` - [EA versions not supported](https://www.home-assistant.io/integrations/unifiprotect/#supported-firmware), Package Camera FPS limitations documented
- [x] `docs-supported-devices` - All [device types documented](https://www.home-assistant.io/integrations/unifiprotect/#supported-devices) (Cameras, Floodlights, Sensors, Viewers, Locks, Chimes, NVR)
- [x] `docs-supported-functions` - All [entities and features](https://www.home-assistant.io/integrations/unifiprotect/#entities) per device type fully documented
- [x] `docs-troubleshooting` - [Troubleshooting section](https://www.home-assistant.io/integrations/unifiprotect/#troubleshooting) covers video delay, audio, liveview, NVR errors
- [x] `docs-use-cases` - Media source, proxy views, and event entities documented with examples
- [x] `dynamic-devices` - [`_async_add_device`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/data.py#L228) dispatches signals for newly adopted devices
- [x] `entity-category` - Entities use `EntityCategory.CONFIG` and `EntityCategory.DIAGNOSTIC` appropriately (e.g., [`sensor.py`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/sensor.py))
- [x] `entity-device-class` - Extensive use of device classes (motion, door, temperature, signal_strength, etc.) in [`binary_sensor.py`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/binary_sensor.py), [`sensor.py`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/sensor.py)
- [x] `entity-disabled-by-default` - Diagnostic and less common entities use `entity_registry_enabled_default=False`
- [x] `entity-translations` - All entities use `translation_key` with translations in [`strings.json`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/strings.json)
- [x] `exception-translations` - Exceptions use `translation_key` for `api_key_required`, `cloud_user`, `protect_version`
- [x] `icon-translations` - Entities use device classes for icons or have static icons that don't require state-based changes.
- [x] `reconfiguration-flow` - async_step_reconfigure implemented in config_flow.py
- [x] `repair-issues` - [`repairs.py`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/repairs.py) implements `RepairsFlow` for cloud_user and RTSP issues
- [x] `stale-devices` - [`_async_remove_device`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/data.py#L247) removes devices from registry when unadopted

## Platinum
- [x] `async-dependency` - [uiprotect library](https://github.com/uilibs/uiprotect) is fully async using aiohttp
- [x] `inject-websession` - Uses [`async_create_clientsession`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/unifiprotect/utils.py#L57) and `async_get_clientsession` for API client
- [x] `strict-typing` - Listed in [`.strict-typing`](https://github.com/home-assistant/core/blob/dev/.strict-typing) file, uses typed `UFPConfigEntry` throughout

---

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [ ] I understand the code I am submitting and can explain how it works.
- [ ] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.
- [ ] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

RaHehl committed:
Add quality scale configuration for UniFi Protect integration

RaHehl committed:
Merge branch 'dev' into unifi-protect-quality-scale

RaHehl committed:
Merge branch 'dev' into unifi-protect-quality-scale

RaHehl committed:
Update quality scale for UniFi Protect integration to platinum

bdraco commented on homeassistant/components/unifiprotect/quality_scale.yaml:
```diff
@@ -0,0 +1,164 @@
+    comment: icons.json defines service icons for all custom actions.
+  reconfiguration-flow:
+    status: done
+    comment: https://github.com/home-assistant/core/pull/157532
```
Can be updated now

bdraco commented:
> Note: The integration currently targets Silver quality scale. The only missing rule for Gold/Platinum is reconfiguration-flow, which is being addressed in a separate PR: https://github.com/home-assistant/core/pull/157532

This is stale above

RaHehl committed:
Update quality_scale.yaml comments based on review feedback

bdraco commented:
Looks good, this one shouldn’t have been too hard since it was Platinum before the quality scale change.

Approved, but left for another member to review as well

joostlek commented on homeassistant/components/unifiprotect/quality_scale.yaml:
```diff
@@ -0,0 +1,164 @@
+rules:
```
`        # _downgrade_ to make ustorage disppear` typo

joostlek commented on homeassistant/components/unifiprotect/quality_scale.yaml:
```diff
@@ -0,0 +1,164 @@
+rules:
```
Or switch + binary sensor

joostlek commented on homeassistant/components/unifiprotect/quality_scale.yaml:
```diff
@@ -0,0 +1,164 @@
+  reauthentication-flow:
+    status: done
+    comment: async_step_reauth and async_step_reauth_confirm implemented in config_flow.py.
+  test-coverage:
```
`json.loads(load_fixture` -> `load_json_object_fixture` or `load_json_array_fixture`

joostlek replied:
nit: `test_reboot_button`, `test_chime_button`, look similar and I think they can be parametrized

joostlek commented on homeassistant/components/unifiprotect/quality_scale.yaml:
```diff
@@ -0,0 +1,164 @@
+  common-modules:
+    status: done
+    comment: Uses data.py for data coordination and entity.py for base entity class.
+  config-flow-test-coverage:
```
I would say that the discovery test that ends in a create entry also needs one to make sure we do that

joostlek commented on homeassistant/components/unifiprotect/quality_scale.yaml:
```diff
@@ -0,0 +1,164 @@
+  common-modules:
+    status: done
+    comment: Uses data.py for data coordination and entity.py for base entity class.
+  config-flow-test-coverage:
```
`test_discovery_with_both_ignored_and_normal_entry` also finish this one

RaHehl replied:
https://github.com/home-assistant/core/pull/159316

joostlek commented on homeassistant/components/unifiprotect/quality_scale.yaml:
```diff
@@ -0,0 +1,164 @@
+  reauthentication-flow:
+    status: done
+    comment: async_step_reauth and async_step_reauth_confirm implemented in config_flow.py.
+  test-coverage:
```
General recommendation to use snapshot tests in the diagnostics test, as it will make sure we have a copy of the data it would expose in the diagnostics

RaHehl replied:
https://github.com/home-assistant/core/pull/159316

joostlek reviewed (changes requested):
Let's put the things we agree to fix in the comments of the rule, and get this merged. I think that we shouldn't add a comment for every rule, only when there's something uncommon. But let's get this merged and then improve from there on

RaHehl committed:
Simplify quality_scale.yaml - remove verbose comments

Remove detailed comments from rules that don't need explanation.
Keep only the exempt rule comment as it explains the exception.

RaHehl committed:
Merge branch 'home-assistant:dev' into unifi-protect-quality-scale

RaHehl committed:
Enhance quality_scale.yaml with todos

RaHehl committed:
Update quality scale to bronze and refine test coverage statuses in quality_scale.yaml

RaHehl committed:
Merge branch 'dev' into unifi-protect-quality-scale

RaHehl commented on homeassistant/components/unifiprotect/quality_scale.yaml:
```diff
@@ -0,0 +1,164 @@
+    comment: BaseProtectEntity sets _attr_has_entity_name = True in entity.py.
+  runtime-data:
+    status: done
+    comment: Uses entry.runtime_data with typed UFPConfigEntry.
```
https://github.com/home-assistant/core/pull/158798

RaHehl committed:
Merge branch 'home-assistant:dev' into unifi-protect-quality-scale

RaHehl committed:
Update quality scale to platinum

RaHehl commented on homeassistant/components/unifiprotect/quality_scale.yaml:
```diff
@@ -0,0 +1,164 @@
+  entity-category:
+    status: done
+    comment: Entities use EntityCategory.CONFIG and EntityCategory.DIAGNOSTIC appropriately.
+  entity-device-class:
```
addressed by entity-device-class comment

RaHehl committed:
Update entity-translations for unifiprotect to include planned improvements

RaHehl commented on homeassistant/components/unifiprotect/quality_scale.yaml:
```diff
@@ -0,0 +1,164 @@
+rules:
```
I would keep it for now and later use it as the basic framework for the doorbell binary sensor deprecation.

RaHehl commented on homeassistant/components/unifiprotect/quality_scale.yaml:
```diff
@@ -0,0 +1,164 @@
+  reauthentication-flow:
+    status: done
+    comment: async_step_reauth and async_step_reauth_confirm implemented in config_flow.py.
+  test-coverage:
```
https://github.com/home-assistant/core/pull/159316

RaHehl commented on homeassistant/components/unifiprotect/quality_scale.yaml:
```diff
@@ -0,0 +1,164 @@
+rules:
```
https://github.com/home-assistant/core/pull/159318

RaHehl committed:
Merge branch 'dev' into unifi-protect-quality-scale

RaHehl commented:
Everything should now be addressed, either via a comment, an already merged PR, or—if they’re optional improvements—at least an open PR.

RaHehl committed:
Merge branch 'dev' into unifi-protect-quality-scale

RaHehl committed:
Merge branch 'dev' into unifi-protect-quality-scale
